### PR TITLE
CoAP: add remote and local endpoint data to coap_pkt_t

### PR DESCRIPTION
**Pull request by fabian18**
### Contribution description

This is an alternative attempt to expose local and remote IP addresses and ports  to the `CoAP` request/response handlers. The `CoAP` PDU now contains pointers to local and remote endpoint data.

### Testing procedure

`examples/gcoap`:

```
> coap get fe80::4c49:6fff:fe54:5c%6 5683 /.well-known/core
2021-09-09 09:42:24,179 # gcoap_cli: sending msg ID 10865, 23 bytes
2021-09-09 09:42:24,183 # gcoap_remote [fe80::4c49:6fff:fe54:5c%6]:5683
2021-09-09 09:42:24,186 # local [fe80::4c49:6fff:fe54:66]:5683
2021-09-09 09:42:24,190 # gcoap: response Success, code 2.05, 46 bytes
2021-09-09 09:42:24,194 # </cli/stats>;ct=0;rt="count";obs,</riot/board>

> coap get fe80::4c49:6fff:fe54:66%6 5683 /.well-known/core
2021-09-09 09:42:25,593 # gcoap_cli: sending msg ID 35428, 23 bytes
2021-09-09 09:42:25,597 # gcoap_remote [fe80::4c49:6fff:fe54:66%6]:5683
2021-09-09 09:42:25,601 # local [fe80::4c49:6fff:fe54:5c]:5683
2021-09-09 09:42:25,605 # gcoap: response Success, code 2.05, 46 bytes
2021-09-09 09:42:25,609 # </cli/stats>;ct=0;rt="count";obs,</riot/board>

```

### Issues/PRs references

Replaces #16429
Trying to solve #15686

### CODE REVIEW

File: sys/net/application_layer/cord/ep/cord_ep.c [hunk 1]
```diff
@@ -91,7 +90,7 @@ static void _on_register(const gcoap_request_memo_t *memo, coap_pkt_t* pdu,
         /* read the location header and save the RD details on success */
         if (coap_get_location_path(pdu, (uint8_t *)_rd_loc,
                                    sizeof(_rd_loc)) > 0) {
-            memcpy(&_rd_remote, remote, sizeof(_rd_remote));
+            memcpy(&_rd_remote, pdu->remote, sizeof(_rd_remote));
```
Comment (maribu) [inline sys/net/application_layer/cord/ep/cord_ep.c:93]: Maybe it makes sense to have a dedicated function for this. This function could `assert()` that the endpoint indeed of a UDP flavor, once multiple protocols are supported.

**Comment by bergzand** [comment]
At a first glance this looks similar to https://github.com/RIOT-OS/RIOT/pull/13621. Did things change enough in the mean time to reconsider adding this?

**Comment by maribu** [comment]
> At a first glance this looks similar to #13621. Did things change enough in the mean time to reconsider adding this?

That was closed in favor of another PR that was closed in the meantime. However, the feature was added by exposing the remote addr as an additional argument. The same could be done here, which is what @fabian18 first did - but that is an API change (PR still open).

I take it that you prefer that version?

**Comment by miri64** [comment]
I think we should at least wait for the outcome of tomorrow's CoAP API breakout session at the summit, before we can conclude on this definitively.

**Comment by bergzand** [comment]
> I take it that you prefer that version?

I'm just passing by and noticed similarities, I haven't looked good enough at all the options to have a good opinion or preference.

**Comment by miri64** [comment]
(see https://hackmd.io/o_y0nKfGRSqrkg-YTCf97g for some initial notes)

**Review by maribu — CHANGES_REQUESTED** [review]
Blocking until breakout session during the summit concludes

**Comment by miri64** [comment]
> Blocking until breakout session during the summit concludes

I think the conclusion was that longterm we move to another API, but short-term [we move out the `gcoap` members out of `nanocoap` into a new `gcoap_pkt_t` struct](https://hackmd.io/o_y0nKfGRSqrkg-YTCf97g#approaches-forward). IMHO, before we worsen things for `nanocoap`, this should be done first, but I don't hard-NACK the approach of merging this first and cleaning up later.

**Pull request closed without merging**